Undo/Redo implementation

YangYue128-helen opened the pull request.

YangYue128-helen committed:
Merge pull request #11 from AY2021S1-CS2103T-T12-4/master

new update

YangYue128-helen committed:
Implement Undo/Redo feature

JinHao-L commented on src/main/java/seedu/address/model/VersionedAppointmentBook.java:
```diff
@@ -0,0 +1,108 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class VersionedAppointmentBook extends AppointmentBook {
```
Might be good to add JavaDocs for the class

JinHao-L commented on src/main/java/seedu/address/model/VersionedPatientBook.java:
```diff
@@ -0,0 +1,108 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class VersionedPatientBook extends PatientBook {
```
Similar to above, might be good to add JavaDocs to the class

JinHao-L commented on src/main/java/seedu/address/model/VersionedAppointmentBook.java:
```diff
@@ -0,0 +1,108 @@
+                && currentStatePointer == versionedAppointmentBook.currentStatePointer;
+    }
+
+    /**
+     * Thrown when trying to {@code undo()} but can't.
+     */
+    public static class NoUndoableStateException extends RuntimeException {
+        private NoUndoableStateException() {
+            super("Current state pointer at start of appointmentBookState list, unable to undo.");
+        }
+    }
+
+    /**
+     * Thrown when trying to {@code redo()} but can't.
+     */
+    public static class NoRedoableStateException extends RuntimeException {
+        private NoRedoableStateException() {
+            super("Current state pointer at end of appointmentBookState list, unable to redo.");
+        }
+    }
```
Might be a bit repetitive since `VersionPatientBook` also has the same exceptions.

JinHao-L reviewed (approved):
Other than the minor comments, the rest LGTM 👍 
The comments can be addressed in a separate PR if necessary

YangYue128-helen committed:
Add javadoc for versionedAppointentBook and versionedPatientBook;
Extract two exceptions out from versionedAppointentBook and versionedPatientBook to avoid repetition.

YangYue128-helen committed:
Merge branch 'master' into Undo-Redo

xz0127 commented on src/main/java/seedu/address/model/VersionedAppointmentBook.java:
```diff
@@ -0,0 +1,98 @@
+    private int currentStatePointer;
+
+    /**
+     * Initialize a versioned appointment book with the given initial state.
```
Perhaps it should be `Initializes`.

xz0127 commented on src/main/java/seedu/address/model/VersionedPatientBook.java:
```diff
@@ -0,0 +1,99 @@
+    private int currentStatePointer;
+
+    /**
+     * Initialize a versioned patient book with the given initial state.
```
Perhaps it should be `Initializes` here as well. 👍

xz0127 commented on src/test/java/seedu/address/model/VersionedAppointmentBookTest.java:
```diff
@@ -0,0 +1,314 @@
+
+public class VersionedAppointmentBookTest {
+
+    private final ReadOnlyAppointmentBook appointmentBookWithAmy = new AppointmentBookBuilder()
```
Would it be better if the name of the appointmentBook is consistent with the patient's name of the appointment? I was a little bit confused when I saw `appointmentBookWithAmy` has appointment `ALICE_APPOINTMENT` added. Perhaps I misunderstood this part. Let me know what you think 👍

xz0127 reviewed (commented):
Other than the minor comments mentioned, the rest LGTM! 👍  Good work and perhaps some additional parts have to be added when `change` and `available` commands are added in.

JinHao-L committed:
Merge branch 'master' into Undo-Redo

JinHao-L committed:
Resolve merge conflict

YangYue128-helen committed:
Rename variables

YangYue128-helen committed:
Merge branch 'master' into Undo-Redo

YangYue128-helen committed:
Now undo/redo supports change command.
